gpkit/nomials/map.py

```diff
@@ -0,0 +1,160 @@
+"Implements the NomialMap class"
+from collections import defaultdict
+import numpy as np
+from .. import DimensionalityError
+from ..small_classes import HashVector, Quantity, Strings
+from ..small_scripts import mag
+from ..varkey import VarKey
+from .substitution import parse_subs
+
+
+DIMLESS_QUANTITY = Quantity(1, "dimensionless")
+
+
+class NomialMap(HashVector):
+    "Class for efficent algebraic represention of a nomial"
+    units = None
+    expmap = None
+    csmap = None
+
+    def set_units(self, thing, thing2=None):
+        "Sets units to those of `thing*thing2`"
+        if thing is None and thing2 is None:
+            self.units = None
+        elif hasattr(thing, "units"):
+            if hasattr(thing2, "units"):
+                self.units = Quantity(1, (thing*thing2).units)
+                try:  # faster than "if self.units.dimensionless"
+                    conversion = float(self.units)
+                    self.units = None
+                    for key, value in self.items():
+                        self[key] = value*conversion
+                except DimensionalityError:
+                    pass
+            else:
+                self.units = Quantity(1, thing.units)
```

Comment:
`thing.units` is a pint.Quantity if `thing` a gpkit object (because pint versions have incompatible uses of the pint.Unit class, we standardize on pint.Quantity) but a pint.Unit if `thing` is a Quantity